gnosis/safe-cmm-app-react
# Feature/#98/add withdraw button back

# Description

Part of #98 

Add withdraw button back (was removed during refactoring).

Withdraw and claim are possible again :)

# To Test
1. Make sure there's at least 1 active strategy
2. Request withdraw for one active strategy

- [ ] Tx to request withdraw should be created

3. After signing the tx, change tabs (at the moment required to update the strategies)

- [ ] Withdraw button should now be disabled and show a countdown for given strategy
- [ ] After countdown is over, withdraw button should be enabled with the text `Claim`

4. Claim your funds for given strategy

- [ ] Tx to claim should be created

5. After signing the tx, change tabs (at the moment required to update the strategies)

- [ ] No request/claim button should be displayed for given strategy

# Background

I avoided doing much rework on this withdraw button logic keeping it mostly the same as the previous version https://github.com/gnosis/safe-cmm-app-react/blob/master/src/routes/Strategies/Active.js#L31

There are some caveats with the current implementation, not sure all of them were there before.
Let's follow step by step:

1. Strategy is active, request button is enabled
![screenshot_2020-10-23_10-11-24](https://user-images.githubusercontent.com/43217/97035683-884d4e00-151b-11eb-81e6-2a6f51277d83.png)

2. Claim has been requested. Transaction pop-us displayed
![screenshot_2020-10-23_10-11-51](https://user-images.githubusercontent.com/43217/97035681-884d4e00-151b-11eb-9bb0-e5539e960be4.png)
  Here, either the user chose to `submit` the transaction or not. We can't tell which from this point forward.
  The button will remain disabled while the screen is active.
  :warning: If the dropdown is closed and re-opened, this state will be lost, and the button will be enabled again

3. If the tx is executed, and the user switches to another tab and back, triggering a refresh on the strategies, the button will show the claim countdown.
![screenshot_2020-10-23_10-13-44](https://user-images.githubusercontent.com/43217/97035680-87b4b780-151b-11eb-94b1-1e275f6db329.png)
  :warning: This state is not triggered automatically because we don't have background strategy update at the moment.

4. Once countdown is over, claim button is enabled
![screenshot_2020-10-23_10-19-27](https://user-images.githubusercontent.com/43217/97035678-87b4b780-151b-11eb-8b3d-a478ba9d5eed.png)

5. If the user had switched to a different screen after the countdown is over, the balances will be reset (notice the amounts at the top on the previous image)
![screenshot_2020-10-23_10-19-55](https://user-images.githubusercontent.com/43217/97035675-871c2100-151b-11eb-9f25-ea14c9fdf296.png)
   :warning: Brackets no longer have balances because withdraw has been requested. We might need a different way to query balances, worth checking how `dex-react` does it

6. Claim has been requested. Transaction pop-us displayed
![screenshot_2020-10-23_10-20-16](https://user-images.githubusercontent.com/43217/97035674-871c2100-151b-11eb-899c-fd72b350f462.png)
  :warning: Same caveats as step `2` applies. If closing dropdown or switching to a different tab state is reset

7. If the tx is executed, and the user switches to another tab and back, triggering a refresh on the strategies, the button will no longer be displayed
![screenshot_2020-10-23_10-20-48](https://user-images.githubusercontent.com/43217/97035669-86838a80-151b-11eb-8504-9748cf1b5b97.png)
  This is a closed strategy, should not be displayed on the active tab

----

In summary, we need background strategy refresh, and further refactor the component to better handle state changes as well as storing more state on global state.
